### PR TITLE
Job sampler

### PR DESCRIPTION
**Pull request by tweddielin**
- Implemented reservoir sampler and weighted reservoir sampler
- Modified `Doc2VecGensimCorpusCreator` to let users select what classes they want to include
- Add `random_state` as an attribute to `JobSampler`
- One thing worth knowing is that in the uni test, I use some statistical characteristics to assert.
    - For reservoir sampling without weighting, I check if the average of each occupation counts equals to total counts divided by occupation numbers after looping for 200 times.
    - For reservoir sampling with weighting, I check if the ratio of the weights falls into the right interval with maximum counts in the histogram as we expect after looping for 200 times.

### CODE REVIEW

File: skills_ml/algorithms/corpus_creators/basic.py [hunk 1]
```diff
@@ -90,11 +90,12 @@ class Doc2VecGensimCorpusCreator(CorpusCreator):
     ]
     join_spaces = ' '.join
 
-    def __init__(self, generator=None):
+    def __init__(self, generator=None, occ_classes=[]):
```
Comment (thcrock) [inline skills_ml/algorithms/corpus_creators/basic.py]: I'd like to see the docstring updated with some information about this new functionality. It already is lacking some information about what the passed-in arguments do, but adding another argument with new functionality makes it even more important to update.

File: skills_ml/algorithms/job_sampler/sampler.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        if i < k:
+            result.append(datum)
+        else:
+            j = random.randint(0, i-1)
+            if j < k:
+                result[j] = datum
+    return result
+
+
+def reservoir_weighted(it, k, weights):
```
Comment (thcrock) [inline skills_ml/algorithms/job_sampler/sampler.py]: Both `reservoir` and `reservoir_weighted` seem pretty generic. Should they just go into an `algorithms.sampling` module? 

This job stuff could even be in a submodule `algorithms.sampling.job`

For that matter, we have a generic `file_sampler` which has similar logic to the unweighted version. Should that be merged with this?

File: skills_ml/algorithms/job_sampler/sampler.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    Args:
+        it (iterator): Job posting iterator to sample from. The format should be (job_posting, label)
+        k (int): Sample size
+        weights (dict): weight
+    """
+    heap = []
+    hkey = lambda w: np.power(np.random.uniform(0.0, 1.0), 1.0 / w)
+    for i, datum in enumerate(it):
+        weight = weights[datum[1]]
```
Comment (thcrock) [inline skills_ml/algorithms/job_sampler/sampler.py]: What expectation is this placing on the input of the function? It looks like it is expecting every label in the iterator to be present as a key in the weights dict. This expectation is something we should specify in the docstring.

File: tests/test_job_sampler.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    js = JobSampler(fake_corpus_train, fake_corpus_train.lookup, weights)
+
+    result = []
+    for i in range(num_loops):
+        r = list(map(lambda x: x[1],js.sample(sample_size)))
+        counts = dict(Counter(r))
+        result.append(counts['13'] / counts['11'])
+
+    hist = np.histogram(result)
+    assert ratio >= hist[1][np.argmax(hist[0])] and ratio <= hist[1][np.argmax(hist[0]) + 1]
```
Comment (thcrock) [inline tests/test_job_sampler.py]: This assertion is pretty advanced, so a comment explaining what's going on would be helpful.

File: skills_ml/algorithms/job_sampler/sampler.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        for job in job_iter:
+            if high_level:
+                yield (job[0], lookup[job[1][0]][:2])
+            else:
+                yield (job[0], lookup[job[1][0]])
+
+    def sample(self, k):
+        it = self._transform_iterator(self.job_posting, self.lookup, self.high_level)
+        if self.weights:
+            return list(reservoir_weighted(it, k, self.weights))
```
Comment (thcrock) [inline skills_ml/algorithms/job_sampler/sampler.py]: Is there a reason for `reservoir` and `reservoir_weighted` to return different things? `reservoir` assumes (correctly, most likely) that the sample itself is small enough to just be a list, and returns the list. `reservoir_weighted`, on the other hand, yields, but its only caller list-ifies its output. I would expect the two base functions to either both return generators or both return lists to make switching between them simpler for a user, if possible.

**Comment by thcrock** [comment]
No big problems, but there are some documentation and organization issues that I think are worth addressing before we merge.

**Comment by tweddielin** [comment]
BTW I want to add more functionalities to corpus_creator to make it more robust and useful in another branch.

**Pull request merged**